network/server/tchannelthrift/pool.go

```diff
@@ -0,0 +1,161 @@
+type BlocksMetadataPool interface {
+	// Get returns a blocks metadata
+	Get() *rpc.BlocksMetadata
+
+	// Put puts a blocks metadata back to pool
+	Put(m *rpc.BlocksMetadata)
+}
+
+// BlocksMetadataSlicePool provides a pool for blocks metadata slices
+type BlocksMetadataSlicePool interface {
+	// Get returns a blocks metadata slice
+	Get() []*rpc.BlocksMetadata
+
+	// Put puts a blocks metadata slice back to pool
+	Put(m []*rpc.BlocksMetadata)
+}
+
+type blockMetadataPool struct {
+	pool pool.ObjectPool
+}
+
+// NewBlockMetadataPool creates a new block metadata pool
+func NewBlockMetadataPool(opts pool.ObjectPoolOptions) BlockMetadataPool {
+	p := pool.NewObjectPool(opts)
+	p.Init(func() interface{} {
+		return rpc.NewBlockMetadata()
+	})
+	return &blockMetadataPool{pool: p}
+}
+
+func (p *blockMetadataPool) Get() *rpc.BlockMetadata {
+	return p.pool.Get().(*rpc.BlockMetadata)
+}
+
+func (p *blockMetadataPool) Put(metadata *rpc.BlockMetadata) {
+	if metadata != nil {
+		*metadata = emptyBlockMetadata
+		p.pool.Put(metadata)
+	}
+}
+
+type blockMetadataSlicePool struct {
+	pool     pool.ObjectPool
+	capacity int
+}
+
+// NewBlockMetadataSlicePool creates a new blockMetadataSlice pool
+func NewBlockMetadataSlicePool(opts pool.ObjectPoolOptions, capacity int) BlockMetadataSlicePool {
+	p := pool.NewObjectPool(opts)
+	p.Init(func() interface{} {
+		return make([]*rpc.BlockMetadata, 0, capacity)
+	})
+	return &blockMetadataSlicePool{pool: p, capacity: capacity}
+}
+
+func (p *blockMetadataSlicePool) Get() []*rpc.BlockMetadata {
+	return p.pool.Get().([]*rpc.BlockMetadata)
+}
+
+func (p *blockMetadataSlicePool) Put(res []*rpc.BlockMetadata) {
+	res = res[:0]
+	p.pool.Put(res)
+}
+
+type blocksMetadataPool struct {
+	pool pool.ObjectPool
+}
+
+// NewBlocksMetadataPool creates a new blocks metadata pool
+func NewBlocksMetadataPool(opts pool.ObjectPoolOptions) BlocksMetadataPool {
+	p := pool.NewObjectPool(opts)
+	p.Init(func() interface{} {
+		return rpc.NewBlocksMetadata()
+	})
+	return &blocksMetadataPool{pool: p}
+}
+
+func (p *blocksMetadataPool) Get() *rpc.BlocksMetadata {
+	return p.pool.Get().(*rpc.BlocksMetadata)
+}
+
+func (p *blocksMetadataPool) Put(metadata *rpc.BlocksMetadata) {
+	if metadata != nil {
+		*metadata = emptyBlocksMetadata
+		p.pool.Put(metadata)
+	}
+}
+
+type blocksMetadataSlicePool struct {
+	pool     pool.ObjectPool
+	capacity int
+}
+
+// NewBlocksMetadataSlicePool creates a new blocksMetadataSlice pool
+func NewBlocksMetadataSlicePool(opts pool.ObjectPoolOptions, capacity int) BlocksMetadataSlicePool {
+	p := pool.NewObjectPool(opts)
+	p.Init(func() interface{} {
+		return make([]*rpc.BlocksMetadata, 0, capacity)
+	})
+	return &blocksMetadataSlicePool{pool: p, capacity: capacity}
+}
+
+func (p *blocksMetadataSlicePool) Get() []*rpc.BlocksMetadata {
+	return p.pool.Get().([]*rpc.BlocksMetadata)
+}
+
+func (p *blocksMetadataSlicePool) Put(res []*rpc.BlocksMetadata) {
+	res = res[:0]
```

Comment:
normally this should just be the fetch batch size but yeah guess doesn't hurt to add it